readme.md
# Plan Config

This Laravel 5 package makes it easy to manage the rules/limits of your SaaS app subscription plans.

## How to install

Pull the package in through Composer.

```js
"require": {
    "seanstewart/plan-config": "dev-master"
}
```

Include the service provider within `app/config/app.php`.

```php
'providers' => [
    Seanstewart\PlanConfig\PlanConfigServiceProvider::class
];
```

Include the facade (optional) in `app/config/app.php`.

```php
'aliases' => [
    'Plan'       => Seanstewart\PlanConfig\Plan::class
];
```

Then you will need to generate your config by running the command

```js
php artisan vendor:publish
```

## How to Use

Let's say your app has subscription plans that limit the number of widgets a user can add. You would have some sort of logic that checks the number of widgets a user is allowed to have in their account. With Plan Config you can do that by calling the helper function plan().

```php
if($this->getCurrentNumberOfWidgets < plan('limits.widgets'))
{
    // Allow the user to add a new widget
}
```

The plan() helper function knows what plan the current user is subscribed to and grabs the limits you defined in your plans.php config file. You can use the helper function anywhere in your application (views, controllers, models, middleware, etc.). Using the previous example, your plan config file would look like this:

```php
'plans' => [

    'bronze'   => [
        'limits' => [
            'widgets' => 5
        ]
    ],

    'silver'   => [
        'limits' => [
            'widgets' => 10
        ]
    ],

    //...and so on

]
```

If your user is subscribed to the silver plan, they could only add 10 widgets. You can even adapt it to use other attributes, like a title, description, or pricing for your plans.

```php
'plans' => [

    'bronze' => [
        'title'       => 'Bronze Plan',
        'description' => 'This is some description for a Bronze Plan',
        'price'       => '19.00',
        'currency'    => 'USD',
        'limits'      => [
            'widgets' => 5
        ]
    ],

    'silver' => [
        'title'       => 'Silver Plan',
        'description' => 'This is some description for the Silver Plan',
        'price'       => '29.00',
        'currency'    => 'USD',
        'limits'      => [
            'widgets' => 10
        ]
    ],

    //...and so on

]
```


#### Get a Plan's Config Without the User

In the case where a specific plan config is needed, you can pass in the plan's code as a string for the 2nd argument:

```
plan('limits.widgets', 'bronze'); // Returns 5
plan('limits.widgets', 'silver'); // Returns 10
```

# Configuring Your Plans

To configure your plans, open up app/plans.php and start adding your plan details. By default the package assumes that you're using laravel's built in Auth, and that the user's plan is stored in the User model. You can set the field used to determine the user's plan in the config...

```php
'plan_field' => 'stripe_plan'
```

To configure your plans, add your plan data in the 'plans' array.

```php
'plans' => [

    'bronze'   => [
        'limits' => [
            'widgets' => 5
        ]
    ],

    'silver'   => [
        'limits' => [
            'widgets' => 10
        ]
    ],

    //...and so on

]
```

If you have rules that apply to all plans, you can define a default or fallback plan. In the config file, set your fallback plan...

```php
'fallback_plan' => '_default',
```

And then define the _default plan in your plans array.

```php
'plans' => [

    '_default' => [
        'limits' => [
            'purple_widgets' => 20
        ]
    ]

    'bronze'   => [
        'limits' => [
            'widgets' => 5
        ]
    ],

    'silver'   => [
        'limits' => [
            'widgets' => 10
        ]
    ],

]
```

In the above example, calling plan('limits.purple_widgets') will give you the value from the fallback plan.

Alternatively you can use the facade and call Plan::get('limits.purple_widgets')

### Overrides

Plan config data can be overridden on the user level by setting an override attribute in the config.

```
'overrides' => [

        // The user model attribute that stores the attributes that can be changed
        'user_model_attribute' => 'plan_overrides',

        // The keys that are allowed to be changed. Set to all by default (['*']).
        'allowed' => ['*'],

    ]

```

In the above example, you would create the *plan_overrides* attribute on the user model. This field should be casted as an array and should include the list of keys and values that should be overridden for the given user.

#### Example

**Plan Config with Overrides:**
```
'overrides' => [

        'user_model_attribute' => 'plan_overrides',

        'allowed' => ['limits.apples', 'limits.bananas'],

    ],


'plans' => [

    '_default' => [
        'limits' => [
            'purple_widgets' => 20
        ]
    ]

    'bronze'   => [
        'limits' => [
            'widgets' => 5,
            'apples' => 10,
            'bananas' => 15
        ]
    ],

    'silver'   => [
        'limits' => [
            'widgets' => 10,
            'apples' => 15,
            'bananas' => 20
        ],
    ],

]
```

**User Model** *plan_overrides* attribute:
```
['limits.widgets' => 1, 'limits.apples' => 50, 'limits.bananas' => 100]
```

Would result in the following...

|  User's Plan | Key | Call | Result | Overridden? |
| --- | --- | --- | --- | --- |
| silver | limits.widgets | `plan('limits.widgets');` | **10** | No |
| silver | limits.apples | `plan('limits.apples');` | **50** | Yes |
| silver | limits.bananas | `plan('limits.bananas');` | **100** | Yes |
| silver | limits.bananas | `plan('limits.bananas', 'silver')` | 20 | No |
| bronze | limits.bananas | `plan('limits.bananas', 'silver')` | 20 | No |
| bronze | limits.bananas | `plan('limits.bananas', $user)` | 100 | Yes |

If you want to return a user's entire plan config (along with overrides), you can pass in `*` as the first argument.

```
plan('*');
```

# Why I created this

I've always found that managing subscriptions and plans for a SaaS app can be complicated. I felt like storing these values in a database isn't the best approach considering a lot of your values and limits will not change frequently. When building [Election Runner](https://electionrunner.com), a web application that allows schools & organizations to run elections, we needed something to accomplish exactly this. Hopefully others will find this as useful as we do!



CHANGELOG.md
#Changelog

All Notable changes to this package` will be documented in this file

## Version 1.4.1

- Allow passing a wildcard `*` as the `$key`  to return the user's entire plan config along with overrides.

## Version 1.4.0

- Significant refactor. The 2nd argument for the `plan()` helper function now allows for passing a $user object.
- Removed the following methods
    - **getPlanOfUser**
    - **getUserPlan**
    - **getAllowedOverrides**
    - **getPlanOverrides**

## Version 1.3.0

### Added

- Added new way to override a plan's config using an attribute on the user model. Requires updating app/config/plans.php with new section (see README).

## Version 1.2.0

### Added

- Unit tests along with travis ci 
- Added optional Facade to use
- Refactored code so its easier to test
- Added new methods

tests/PlanConfigTest.php
<?php

use Illuminate\Contracts\Foundation\Application;
use Illuminate\Support\Facades\App;
use Illuminate\Support\Facades\Config;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\Facade;
use Mockery as m;
use Orchestra\Testbench\TestCase;
use Seanstewart\PlanConfig\Plan;
use Seanstewart\PlanConfig\PlanConfig;

class PlanConfigTest extends TestCase {

    protected $planConfig;
    protected $config;
    protected $auth;

    public function setUp()
    {
        parent::setUp();
        $this->config = new Config();
        $this->auth = new Auth();
        $this->planConfig = new Seanstewart\PlanConfig\PlanConfig($this->config, $this->auth);
    }

    protected function getPackageProviders($app)
    {
        return [\Seanstewart\PlanConfig\PlanConfigServiceProvider::class];
    }

    protected function getPackageAliases($app)
    {
        return [
            'Plan' => Plan::class
        ];
    }

    /**
     * @test
     * @group function
     */
    public function it_tests_global_function_plan()
    {
        Plan::shouldReceive('get')
            ->once()
            ->with('key', 'plan')
            ->andReturn('foo!');

        $this->assertEquals('foo!', plan('key', 'plan'));

        Plan::shouldReceive('get')
            ->once()
            ->with('key', null)
            ->andReturn('foo!');

        $this->assertEquals('foo!', plan('key'));
    }

    /**
     * @test
     * @group function
     */
    public function it_tests_global_function_plan_no_key()
    {
        Plan::shouldReceive('get')->never();

        $this->assertInstanceOf(Seanstewart\PlanConfig\PlanConfig::class, plan(null));
    }

    /**
     * @test
     */
    public function it_tests_get_with_null_for_plan()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['setContext', 'getCurrentUserPlan', 'getPlanKey'], [$this->config, $this->auth]);

        $planConfig->expects($this->once())
                   ->method('setContext')
                   ->with(null);

        $planConfig->expects($this->once())
                   ->method('getCurrentUserPlan')
                   ->willReturn('current_user_plan');

        $planConfig->expects($this->once())
                   ->method('getPlanKey')
                   ->with('key', 'current_user_plan')
                   ->willReturn(true);

        $this->assertTrue($planConfig->get('key'));
    }

    /**
     * @test
     */
    public function it_tests_get_with_plan()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['setContext', 'getCurrentUserPlan', 'getPlanKey'], [$this->config, $this->auth]);

        $planConfig->expects($this->never())
                   ->method('setContext');

        $planConfig->expects($this->never())
                   ->method('getCurrentUserPlan');

        $planConfig->expects($this->once())
                   ->method('getPlanKey')
                   ->with('key', 'userPlan')
                   ->willReturn(true);

        $this->assertTrue($planConfig->get('key', 'userPlan'));
    }

    /**
     * @test
     */
    public function it_tests_get_with_user()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['setContext', 'getCurrentUserPlan', 'getPlanKey'], [$this->config, $this->auth]);

        $user = ['id' => 123];

        $planConfig->expects($this->once())
                   ->method('setContext')
                   ->with($user);

        $planConfig->expects($this->once())
                   ->method('getCurrentUserPlan')
                   ->willReturn('current_user_plan');

        $planConfig->expects($this->once())
                   ->method('getPlanKey')
                   ->with('key', 'current_user_plan')
                   ->willReturn(true);

        $this->assertTrue($planConfig->get('key', $user));
    }

    /**
     * @test
     */
    public function it_tests_get_all_plans()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getFallbackPlan'], [$this->config, $this->auth]);

        $plans = [
            'plan_1' => ['foo1' => 'bar1'],
            'plan_2' => ['foo2' => 'bar2'],
        ];

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('plans')
                   ->willReturn($plans);

        $this->assertEquals($plans, $planConfig->getAllPlans($plans));
    }

    /**
     * @test
     */
    public function it_gets_plan()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getFallbackPlan'], [$this->config, $this->auth]);

        $fallbackPlan = ['bar' => 'foo'];

        $planGroup1 = [
            'plan_1' => ['foo1' => 'bar1'],
            'plan_2' => ['foo2' => 'bar2'],
        ];

        $planGroup2 = [
            'plan_3' => ['foo3' => 'bar3']
        ];

        $planConfig->expects($this->exactly(2))
                   ->method('getConfig')
                   ->withConsecutive(['plans', null], ['plans', null])
                   ->willReturnOnConsecutiveCalls($planGroup1, $planGroup2);

        $planConfig->expects($this->exactly(2))
            ->method('getFallbackPlan')
            ->willReturnOnConsecutiveCalls($fallbackPlan, $fallbackPlan);

        $this->assertEquals($planGroup1['plan_1'], $planConfig->getPlan('plan_1'));
        $this->assertEquals($fallbackPlan, $planConfig->getPlan('plan_2'));
    }

    /**
     * @test
     */
    public function it_gets_fallback_plan()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getAllPlans'], [$this->config, $this->auth]);

        $plans = [
            'default' => ['bar' => 'foo'],
            'plan_1' => ['foo1' => 'bar1'],
            'plan_2' => ['foo2' => 'bar2']
        ];

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('fallback_plan')
                   ->willReturn('default');

        $planConfig->expects($this->once())
                   ->method('getAllPlans')
                   ->willReturn($plans);

        $this->assertEquals($plans['default'], $planConfig->getFallbackPlan());
    }

    /**
     * @test
     */
    public function it_gets_fallback_plan_no_fallback()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getAllPlans'], [$this->config, $this->auth]);

        $plans = [
            'plan_1' => ['foo1' => 'bar1'],
            'plan_2' => ['foo2' => 'bar2'],
        ];

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('fallback_plan')
                   ->willReturn('default');

        $planConfig->expects($this->once())
                   ->method('getAllPlans')
                   ->willReturn($plans);

        $this->assertEquals([], $planConfig->getFallbackPlan());
    }

    /**
     * @test
     */
    public function it_gets_config()
    {
        $config = [
            'plans' => [
                'foo' => [
                    'bar' => 'foobar'
                ]
            ]
        ];

        $this->config->shouldReceive('get')
                     ->with('plans')
                     ->andReturn($config['plans']);

        $this->assertEquals($config['plans'], $this->planConfig->getConfig());
        $this->assertEquals('foobar', $this->planConfig->getConfig('foo.bar'));
        $this->assertEquals(null, $this->planConfig->getConfig('foo.bar2'));
        $this->assertEquals('default', $this->planConfig->getConfig('foo.bar2', 'default'));
    }

    /**
     * @test
     */
    public function it_tests_get_plan_key()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getPlan', 'getAllowedOverrides'], [$this->config, $this->auth]);

        $completePlan = ['foo' => ['bar' => 'value'], 'barfoo' => 'abc123'];

        $planConfig->expects($this->exactly(4))
                   ->method('getPlan')
                   ->with('plan')
                   ->willReturnOnConsecutiveCalls(
                       ['foo' => ['bar' => 'value']],
                       ['foo' => 'bar'],
                       ['foo' => ['bar' => 'value'], 'barfoo' => 'abc123'],
                       $completePlan
                   );

        $planConfig->expects($this->exactly(4))
                   ->method('getAllowedOverrides')
                   ->willReturnOnConsecutiveCalls([], [], ['barfoo' => 'barfoobarfoo'], ['barfoo' => 'barfoobarfoo']);

        $this->assertEquals('value', $planConfig->getPlanKey('foo.bar', 'plan'));
        $this->assertNull($planConfig->getPlanKey('foo.bar', 'plan'));
        $this->assertEquals('barfoobarfoo', $planConfig->getPlanKey('barfoo', 'plan'));
        $this->assertEquals([
            'foo.bar' => 'value',
            'barfoo' => 'barfoobarfoo'
        ], $planConfig->getPlanKey('*', 'plan'));

    }

    /**
     * @test
     */
    public function it_tests_get_plan_key_with_overrides()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getPlan', 'getAllowedOverrides'], [$this->config, $this->auth]);

        $planConfig->expects($this->exactly(2))
                   ->method('getPlan')
                   ->with('plan')
                   ->willReturnOnConsecutiveCalls(['foo' => ['bar' => 'value']], ['foo' => 'bar']);

        $planConfig->expects($this->exactly(2))
                   ->method('getAllowedOverrides')
                   ->willReturnOnConsecutiveCalls(['foo.bar' => 'override', 'extra.key' => 'foo'], []);

        $this->assertEquals('override', $planConfig->getPlanKey('foo.bar', 'plan'));
        $this->assertEquals('bar', $planConfig->getPlanKey('foo', 'plan'));
    }

    /**
     * @test
     */
    public function it_tests_extract_user_plan_overrides()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig'], [$this->config, $this->auth]);

        $overrides = ['foo.bar' => 'foo'];

        $user1 = [];
        $user2 = ['plan_overrides' => $overrides];

        $planConfig->expects($this->exactly(2))
                   ->method('getConfig')
                   ->with('overrides.user_model_attribute', null)
                   ->willReturn('plan_overrides');

        $this->assertEquals([], $planConfig->extractUserPlanOverrides($user1));
        $this->assertEquals($overrides, $planConfig->extractUserPlanOverrides($user2));
    }

    /**
     * @test
     */
    public function it_tests_get_allowed_overrides_no_overrides()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getCurrentUserPlanOverrides'], [$this->config, $this->auth]);

        $overrides = null;

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('overrides.allowed')
                   ->willReturn(null);

        $planConfig->expects($this->once())
                   ->method('getCurrentUserPlanOverrides')
                   ->willReturn([]);

        $this->assertEquals([], $planConfig->getAllowedOverrides('plan_overrides'));
    }

    /**
     * @test
     */
    public function it_tests_get_allowed_overrides_key_not_set()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getCurrentUserPlanOverrides'], [$this->config, $this->auth]);

        $overrides = ['foo' => 'bar'];

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('overrides.allowed')
                   ->willReturn(null);

        $planConfig->expects($this->once())
                   ->method('getCurrentUserPlanOverrides')
                   ->willReturn($overrides);

        $this->assertEquals([], $planConfig->getAllowedOverrides('plan_overrides'));
    }

    /**
     * @test
     */
    public function it_tests_get_allowed_overrides_allows_all_keys()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getCurrentUserPlanOverrides'], [$this->config, $this->auth]);

        $overrides = ['foo' => ['bar' => 'foobar']];

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('overrides.allowed')
                   ->willReturn(['*']);

        $planConfig->expects($this->once())
                   ->method('getCurrentUserPlanOverrides')
                   ->willReturn($overrides);

        $this->assertEquals(['foo.bar' => 'foobar'], $planConfig->getAllowedOverrides('plan_overrides'));
    }

    /**
     * @test
     */
    public function it_tests_get_allowed_overrides_returns_allowed_only()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig', 'getCurrentUserPlanOverrides'], [$this->config, $this->auth]);

        $overrides = [
            'foo'          => ['bar' => 'foobar'],
            'yes'          => ['foobar' => 'yeah'],
            'foobarfoobar' => 123,
            'nooverride'   => 'foo'
        ];

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('overrides.allowed')
                   ->willReturn(['foo.bar', 'yes.foobar', 'foobarfoobar']);

        $planConfig->expects($this->once())
                   ->method('getCurrentUserPlanOverrides')
                   ->willReturn($overrides);

        $this->assertEquals([
            'foo.bar'      => 'foobar',
            'yes.foobar'   => 'yeah',
            'foobarfoobar' => 123
        ], $planConfig->getAllowedOverrides('plan_overrides'));
    }

    /**
     * @test
     */
    public function it_tests_extract_user_plan()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getConfig'], [$this->config, $this->auth]);

        $planConfig->expects($this->once())
                   ->method('getConfig')
                   ->with('plan_field')
                   ->willReturn('stripe_plan');

        $user = [
            'stripe_plan' => 'foobar'
        ];

        $this->assertEquals('foobar', $planConfig->extractUserPlan($user));
    }

    /**
     * @test
     */
    public function it_tests_get_authenticated_user()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getPlan'], [$this->config, $this->auth]);

        Auth::shouldReceive('check')
            ->once()
            ->andReturn(true);

        $user = ['id' => 123];

        Auth::shouldReceive('getUser')
            ->once()
            ->andReturn($user);

        $this->assertEquals($user, $planConfig->getAuthenticatedUser());
    }

    /**
     * @test
     */
    public function it_tests_get_authenticated_user_not_logged_in()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getPlan'], [$this->config, $this->auth]);

        Auth::shouldReceive('check')
            ->andReturn(false);

        Auth::shouldReceive('getUser')->never();

        $this->assertEquals([], $planConfig->getAuthenticatedUser());

    }

    /**
     * @test
     */
    public function it_tests_set_context()
    {
        $planConfig = $this->getMock('SeanStewart\PlanConfig\PlanConfig', ['getAuthenticatedUser', 'extractUserPlan', 'extractUserPlanOverrides'], [$this->config, $this->auth]);

        $user = ['id' => 1];

        $planConfig->expects($this->never())
                   ->method('getAuthenticatedUser');

        $planConfig->expects($this->once())
                   ->method('extractUserPlan')
                   ->with($user)
                   ->willReturn('plan');

        $planConfig->expects($this->once())
                   ->method('extractUserPlanOverrides')
                   ->with($user)
                   ->willReturn(['overrides']);

        $planConfig->setContext($user);

        $this->assertEquals('plan', $planConfig->getCurrentUserPlan());
        $this->assertEquals(['overrides'], $planConfig->getCurrentUserPlanOverrides());
    }
}

src/PlanConfig.php
<?php
/**
 * Created by PhpStorm.
 * User: seanstewart
 * Date: 8/9/15
 * Time: 7:58 AM
 */

namespace Seanstewart\PlanConfig;

use Illuminate\Support\Arr;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\Config;

class PlanConfig {

    public $config;

    /**
     * @var string
     */
    protected $fallbackPlan = [];

    /**
     * @var
     */
    protected $configPlanField;

    /**
     * @var Auth
     */
    protected $auth;

    /**
     * @var
     */
    protected $currentUserPlanOverrides = [];

    /**
     * @var
     */
    protected $currentUserPlan;

    /**
     * PlanConfig constructor.
     * @param Config $config
     * @param Auth $auth
     */
    function __construct(Config $config, Auth $auth)
    {
        $this->auth = $auth;
        $this->config = $config;
    }

    /**
     * Get the config key
     * @param $key
     * @param string|object|array $plan
     * @return bool
     */
    public function get($key, $plan = null)
    {
        // If a string is provided, then we want to look up the key for a given plan
        if (is_string($plan))
        {
            return $this->getPlanKey($key, $plan);
        }

        // If the provided plan is not a string, then we are looking up a plan for a specific user and not
        // the actual config for a specific plan. This allows us to factor in the user's overrides
        $this->setContext($plan);

        return $this->getPlanKey($key, $this->getCurrentUserPlan());
    }

    /**
     * Returns all plans
     * @return mixed
     */
    public function getAllPlans()
    {
        return $this->getConfig('plans');
    }

    /**
     * @param $plan
     * @return mixed
     */
    public function getPlan($plan)
    {
        $plans = $this->getConfig('plans');

        return Arr::get($plans, $plan, $this->getFallbackPlan());
    }

    /**
     * @param $key
     * @param $plan
     * @return bool
     */
    public function getPlanKey($key, $plan)
    {
        $plan = $this->getPlan($plan);

        // Since the key should be sent as array_dot
        // we need to convert the plan array to array_dot
        $planDot = Arr::dot($plan);

        // Merge the plan data with the overrides
        $planDot = array_merge($planDot, $this->getAllowedOverrides());

        if($key == '*')
        {
            return $planDot;
        }

        return Arr::get($planDot, $key, null);
    }

    /**
     * Returns the default fallback plan
     * @return mixed
     */
    public function getFallbackPlan()
    {
        return Arr::get($this->getAllPlans(), $this->getConfig('fallback_plan'), []);
    }

    /**
     * Returns the config for this package.
     * If a key is provided, then the value of the key will be returned.
     * If a key is provided and it doesn't exist, then the default value will be returned.
     * will be returned.
     * @param null $key
     * @param null $default
     * @return mixed
     */
    public function getConfig($key = null, $default = null)
    {
        $config = Config::get('plans');

        return $key ? Arr::get($config, $key, $default) : $config;
    }

    /**
     * Returns the user's plan overrides with only the keys allowed from the config
     * @return array
     */
    public function getAllowedOverrides()
    {
        $allowed = $this->getConfig('overrides.allowed');

        $overrides = $this->getCurrentUserPlanOverrides();

        // If we are allowing all overrides, then return all
        if ($allowed == ['*'])
        {
            return Arr::dot($overrides);
        }

        // Only return the overrides that are allowed
        return Arr::only(Arr::dot($overrides), $allowed);
    }

    /**
     * Returns the user's plan overrides
     * @param $user
     * @return array
     */
    public function extractUserPlanOverrides($user)
    {
        $attribute = $this->getConfig('overrides.user_model_attribute');

        // If this value is null, return an empty array
        if (!$attribute)
        {
            return [];
        }

        // Get the overrides from the user. If none, return an empty array.
        return Arr::get($user, $attribute, []);
    }

    /**
     * Returns the user's plan
     * @param $user
     * @return string
     */
    public function extractUserPlan($user)
    {
        $modelPlanAttribute = $this->getConfig('plan_field');

        return Arr::get($user, $modelPlanAttribute);
    }

    /**
     * @return array
     */
    public function getAuthenticatedUser()
    {
        return Auth::check() ? Auth::getUser() : [];
    }

    /**
     * Sets the user context for looking up the plan
     * @param $user
     */
    public function setContext($user)
    {
        // If a user isn't provided, then we want to get the authenticated user
        if(!$user)
        {
            $user = $this->getAuthenticatedUser();
        }

        $this->currentUserPlan = $this->extractUserPlan($user);
        $this->currentUserPlanOverrides = $this->extractUserPlanOverrides($user);
    }

    /**
     * Provides the current user's plan
     * @return mixed
     */
    public function getCurrentUserPlan()
    {
        return $this->currentUserPlan;
    }

    /**
     * Provides the current user's overrides
     * @return array
     */
    public function getCurrentUserPlanOverrides()
    {
        return $this->currentUserPlanOverrides;
    }

}

src/plans.php
<?php

return [

    /*
   |--------------------------------------------------------------------------
   | User Model Plan Field
   |--------------------------------------------------------------------------
   |
   | Define the field that is used on the User model to define the
   | plan that the user is subscribed to.
   */

    'plan_field' => 'stripe_plan',


    /*
   |--------------------------------------------------------------------------
   | Fallback Plan
   |--------------------------------------------------------------------------
   |
   | The fallback plan will be used if one of the requested attributes
   | is not found in the user's subscription plan. If you don't define a
   | default fallback plan, then set this to false.
   */

    'fallback_plan' => '_default',

    /*
	|--------------------------------------------------------------------------
	| Overrides
	|--------------------------------------------------------------------------
	|
	| Here you can specify which keys can have their values changed using a model attribute
	*/

    'overrides' => [

        // The user model attribute that stores the attributes that can be changed
        'user_model_attribute' => 'plan_overrides',

        // The keys that are allowed to be changed. Set to all by default (['*']).
        'allowed' => ['*'],

    ],

    /*
	|--------------------------------------------------------------------------
	| Subscription plans
	|--------------------------------------------------------------------------
	|
	| Here you may define all of the plans that you offer, along with the
    | limits for each plan. The default plan will be used if no plan matches
	| the one provided when calling the plan() helper function.
	*/

    'plans'      => [

        '_default' => [

        ],

        'bronze' => [
            'title'       => 'Bronze Plan',
            'description' => 'This is some description for a Bronze Plan',
            'price'       => '19.00',
            'currency'    => 'USD',
            'limits'      => [
                'widgets' => 5
            ]
        ],

        'silver' => [
            'title'       => 'Silver Plan',
            'description' => 'This is some description for the Silver Plan',
            'price'       => '29.00',
            'currency'    => 'USD',
            'limits'      => [
                'widgets' => 10
            ]
        ],

        //...and so on

    ]
];
